.github/workflows/ci.yml

```diff
@@ -0,0 +1,275 @@
+name: Gazebo CI
+
+on:
+  push:
+    tags:
+      - self-hosted-*
+    branches:
+      - main
+      - staging
+  pull_request:
+
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
+  cancel-in-progress: true
+
+env:
+  total-runners: 6
+
+jobs:
+  install:
+    name: Install deps
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - uses: actions/setup-node@v3
+        with:
+          node-version: '18.8.0'
+      - name: Cache node_modules
+        id: cache-node-modules
+        uses: actions/cache@v3
+        env:
+          cache-name: cache-gazebo-node-modules
+        with:
+          path: |
+            node_modules
+          key: ${{ runner.os }}-${{ env.cache-name }}-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-${{ env.cache-name }}-
+
+      - name: Install packages
+        if: ${{ steps.cache-node-modules.outputs.cache-hit != 'true' }}
+        run: |
+          npm ci
+  build:
+    name: Build App
+    runs-on: ubuntu-latest
+    needs: [install]
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+      - uses: actions/setup-node@v3
+        with:
+          node-version: '18.8.0'
+      - name: Cache node_modules
+        id: cache-node-modules
+        uses: actions/cache@v3
+        env:
+          cache-name: cache-gazebo-node-modules
+        with:
+          path: |
+            node_modules
+          key: ${{ runner.os }}-${{ env.cache-name }}-${{ hashFiles('**/package-lock.json') }}
+          restore-keys: |
+            ${{ runner.os }}-${{ env.cache-name }}-
+
+      - name: Install packages
+        if: ${{ steps.cache-node-modules.outputs.cache-hit != 'true' }}
+        run: |
+          npm run build
```

Comment:
Think may be a bit of copy paste error here, by this step we should have `node_modules` present so you probably don't need to cache them again, as well you should probably always be good to `npm run build` without a check for a cache miss. As well the last step here probably should be named `Build React App`

Reply:
I'll test it out, but my understanding is that each of these steps runs in isolation, so for node_modules to exist, I need to call the cache step. I'm using cache less as an ephemeral cache and more as a persistent store between jobs.